### ISSUE TITLE
JOSESwift jwe encryption failed to decode in the nimbus server

### ISSUE THREAD

**Issue by ashrafulkarim** [airsidemobile/JOSESwift#112]
Had anybody used JOSESwift successfully? In my case, decryption in the server failing, probably cannot find the matching private key or wrong with encryption. Getting error 500.

My code is, getting the public keys from a server.

```
keys?.keys?.forEach({ (key) in
                BPLogger.debug("\(key)")
                do {
                    let jwkData = key.toJSONString()?.data(using: .utf8)
                    let rsaKey = try RSAPublicKey(data: jwkData!)
                    BPLogger.log("key components: \(rsaKey.parameters)")
                    BpidCache.shared.joseRsaKey = rsaKey
                    self?.generateParametersJose()
                    completion()                        
                    return
                } catch  {
                    BPLogger.debug("Error: \(error)")
                }
            })
```
The server expected a 'kid' field in the jose header, which was missing in the framework. So I have added it... The backend Java server uses nimbus library.

 ```
func generateParametersJose() {
        let rsa = BpidCache.shared.joseRsaKey
        var publicKey: SecKey? = nil
        do {
            publicKey = try rsa?.converted(to: SecKey.self)
        } catch {
            BPLogger.log("\(error)")
        }
        var header = JWEHeader(algorithm: .RSA1_5, encryptionAlgorithm: .A256CBCHS512)
//      header.parameters["kid"] = "1"
        let jwk = MidApi.Model.JWTKey(key: cek);
        let jwkData = try! JSONEncoder().encode(jwk)
        BPLogger.debug("jwkData = \(String(data: jwkData, encoding: .utf8)!)")
        let payload = Payload(jwkData)
        // Encrypter algorithms must match header algorithms.
        guard let encrypter = Encrypter<SecKey>(keyEncryptionAlgorithm: .RSA1_5, encryptionKey: publicKey!, contentEncyptionAlgorithm: .A256CBCHS512) else {
            return
        }
        guard let jwe = try? JWE(header: header, payload: payload, encrypter: encrypter) else {
            BPLogger.error("Falied jwe creation.")
            return
        }

        var comps = jwe.compactSerializedString.components(separatedBy: ".")
        var jweHeader = comps.first
        let data = jweHeader?.base64URLDecode()
        var orgH = try! JSONDecoder().decode(BPJweHeader.self, from: data!)
        orgH.kid = "1"
        let newJson = try! JSONEncoder().encode(orgH).base64URLEncodedString()
        comps[0] = newJson
        let newHeader = comps.joined(separator: ".")
        BPLogger.log("jwe.compactSerializedString = \(newHeader)")
        headers = ["X-Encrypted-Key": newHeader]
//      headers = ["X-Encrypted-Key": jwe.compactSerializedString] // this also fails
    }
```

What am I doing wrong?
corresponding so question, https://stackoverflow.com/questions/52643217/joseswift-jwe-encryption-failed-to-decode-in-the-nimbus-server

**Comment by daniel-mohemian** [comment]
Hi, @ashrafulkarim 👋 

It looks like you're changing the header after encrypting the JWE. Because the header is used as additional authenticated data when encrypting the JWE, changing the header afterwards invalidates the encrypted data. 

To make it work you have to configure your header before using it in a JWE. Unfortunately, it is currently not possible to set additional parameters in headers due to a bug which is currently being fixed in #110. Expect this pull request being merged in the next couple of days!

Until then you should be able to work around this issue by changing [this initializer](https://github.com/airsidemobile/JOSESwift/blob/master/JOSESwift/Sources/JWEHeader.swift#L58) in [`JWEHeader.swift`](https://github.com/airsidemobile/JOSESwift/blob/master/JOSESwift/Sources/JWEHeader.swift) to:

``` swift
/// Initializes a `JWEHeader` with the specified algorithm and signing algorithm.
public init(algorithm: AsymmetricKeyAlgorithm, encryptionAlgorithm: SymmetricKeyAlgorithm, kid: String? = nil) {
    let parameters = [
        "alg": algorithm.rawValue,
        "enc": encryptionAlgorithm.rawValue
    ]

    if let kid = kid {
        parameters["kid"] = kid
    }

    // Forcing the try is ok here, since [String: String] can be converted to JSON.
    // swiftlint:disable:next force_try
    let headerData = try! JSONSerialization.data(withJSONObject: parameters, options: [])

    // Forcing the try is ok here, since "alg" and "enc" are the only required header parameters.
    // swiftlint:disable:next force_try
    try! self.init(parameters: parameters, headerData: headerData)
}
```

**Comment by ashrafulkarim** [comment]
Thanks for the tips. And Thanks again for the lib. I have created a local pod to add the 'kid' param and it worked.

**Comment by daniel-mohemian** [comment]
Good to hear. Again, keep an eye on #110 which will enable additional parameters in the header.

**Comment by daniel-mohemian** [comment]
This is now fixed via #110. See [readme](https://github.com/airsidemobile/JOSESwift#header-1) or [wiki](https://github.com/airsidemobile/JOSESwift/wiki/JWE#header-1) for details.
